components/cldr-json-data-provider/src/transform/dates.rs

```diff
@@ -0,0 +1,368 @@
+use crate::cldr_langid::CldrLangID;
+use crate::error::Error;
+use crate::reader::{get_subdirectories, open_reader};
+use crate::support::DataKeySupport;
+use crate::CldrPaths;
+use icu_data_provider::iter::DataEntryCollection;
+use icu_data_provider::prelude::*;
+use icu_data_provider::structs::dates::*;
+use std::convert::TryFrom;
+use std::marker::PhantomData;
+
+/// A data provider reading from CLDR JSON dates files.
+#[derive(PartialEq, Debug)]
+pub struct DatesProvider<'d> {
+    data: Vec<(CldrLangID, cldr_json::LangDates)>,
+    _phantom: PhantomData<&'d ()>, // placeholder for when we need the lifetime param
+}
+
+impl TryFrom<&CldrPaths> for DatesProvider<'_> {
+    type Error = Error;
+    fn try_from(cldr_paths: &CldrPaths) -> Result<Self, Self::Error> {
+        let mut data = vec![];
+
+        let path = cldr_paths.cldr_dates.clone()?.join("main");
+
+        let locale_dirs = get_subdirectories(&path).unwrap();
+
+        for dir in locale_dirs {
+            let path = dir.join("ca-gregorian.json");
+
+            let langid: CldrLangID = dir.file_name().unwrap().to_str().unwrap().parse().unwrap();
+            let resource: cldr_json::Resource =
+                serde_json::from_reader(open_reader(path)?).unwrap();
+            let lang_dates = resource
+                .main
+                .0
+                .into_iter()
+                .find(|(lid, _)| lid == &langid)
+                .map(|(_, val)| val)
+                .unwrap();
+            match data.binary_search_by_key(&&langid, |(langid, _)| langid) {
```

Comment:
FYI, CldrLangID is crafted such that if you read the locales in lexigraphical order, you don't need to re-sort them (including with regard to the root->und mapping).

Reply:
Quoting https://doc.rust-lang.org/std/fs/fn.read_dir.html
```
The order in which this iterator returns entries is platform and filesystem dependent.
```

I'd prefer not to assume the order.

Reply:
Got it.  I didn't know that `fs::read_dir` didn't return stuff in lexicographical order.

Optional: since you already have a helper function `get_subdirectories`, consider performing the sorting inside of that function, such that you can carry that assumption out to here and simplify the code.